### PR TITLE
SBML etree

### PR DESCRIPTION
**Pull request by DavAug**
- Minidom has been replaced by ElementTree. 
- Tests have been extended to check the correctness of the myokit model, defined by the SBML file.
- A flag has been introduced to add a time bound variable to the myokit model (not automatically included in the SBML definition, but implicitly assumed)

The tests do not yet include different SBML versions (only level 2, version 3) because that is the version of the Hodgkin Huxley model. I am planning to add a test for each of the supported versions in the next pull request, when we include compartment, species and reaction support. ( The TwoComp.xml will be there for that purpose)

### CODE REVIEW

File: myokit/formats/sbml/_importer.py [hunk 1]
```diff
@@ -96,53 +102,76 @@ def model(self, path):
...
-        x = dom_child(xmodel, 'listOfUnitDefinitions')
+        x = xmodel.find(self.ns['sbml'] + 'listOfUnitDefinitions')
         if x:
-            self._parse_units(model, comp, x)
+            self._parse_units(model, comp, x, self.ns['sbml'])
+
+        # add time as independent variable (not explicit in SBML format)
+        if bind_time:
+            # add and bind time variable to component
+            comp.add_variable('time')
```
Comment (MichaelClerx) [inline myokit/formats/sbml/_importer.py:143]: What happens if there's already a variable called 'time' ?

File: myokit/_sim/fiber_tissue.py [hunk 2]
```diff
@@ -365,17 +365,20 @@ def _create_backend(self):
 
         # Define libraries
         libs = []
+        flags = []
```
Comment (MichaelClerx) [inline myokit/_sim/fiber_tissue.py:368]: I think I've managed to fix the OpenCL issues on os/x, but in a separate PR. We'll have to find some way to revert the changes to these three files before we can merge this PR

File: myokit/formats/sbml/_importer.py [hunk 3]
```diff
@@ -67,21 +66,28 @@ def _convert_unit(self, unit):
     def info(self):
         return info
 
-    def model(self, path):
-
+    def model(self, path, bind_time=True):
+        """
+        Returns myokit model specified by the SBML fil provided.
```
Comment (MichaelClerx) [inline myokit/formats/sbml/_importer.py:71]: Maybe some like "Returns a :class:`myokit.Model` based on the SBML file provided.

Also would need some explanation of what `bind_time` does

File: myokit/formats/sbml/_importer.py [hunk 4]
```diff
@@ -96,53 +102,76 @@ def model(self, path):
...
+            comp.add_variable('time')
+            time = comp.get('time')
+            time.set_binding('time')
+            # set unit and value
+            if 'time' in self.units.keys():
+                unit = self.units['time']
+            else:
+                unit = myokit.units.s
+                log.warn('Unit of time could not be found in file and was by '
+                         + 'default set to seconds.')
```
Comment (MichaelClerx) [inline myokit/formats/sbml/_importer.py:152]: You can actually remove the + here and it'll still work!

File: myokit/formats/sbml/_importer.py [hunk 5]
```diff
@@ -96,53 +102,76 @@ def model(self, path):
...
+            self._parse_units(model, comp, x, self.ns['sbml'])
+
+        # add time as independent variable (not explicit in SBML format)
+        if bind_time:
+            # add and bind time variable to component
+            comp.add_variable('time')
+            time = comp.get('time')
+            time.set_binding('time')
+            # set unit and value
+            if 'time' in self.units.keys():
```
Comment (MichaelClerx) [inline myokit/formats/sbml/_importer.py:147]: A slightly more efficient way to do this is:
```
try:
    unit = self.units['time']
except KeyError:
    unit = myokit.units.s
    log.warn(...)
```
This version only has a single dict lookup (and some python fans prefer it for more philosophical reasons too)

File: myokit/formats/sbml/_importer.py [hunk 6]
```diff
@@ -96,53 +102,76 @@ def model(self, path):
         refs = {}
 
         # Handle notes, if given
-        x = dom_child(xmodel, 'notes')
+        x = xmodel.find(self.ns['sbml'] + 'notes')
```
Comment (MichaelClerx) [inline myokit/formats/sbml/_importer.py:105]: This won't detect if there's two 'notes' elements, but I suppose we're not super interested in detecting invalid files here...

File: myokit/formats/sbml/_importer.py [hunk 7]
```diff
@@ -96,53 +102,76 @@ def model(self, path):
...
+        x = xmodel.find(self.ns['sbml'] + 'listOfFunctionDefinitions')
         if x:   # pragma: no cover
             log.warn('Custom math functions are not (yet) implemented.')
 
         # Parse custom units
-        x = dom_child(xmodel, 'listOfUnitDefinitions')
+        x = xmodel.find(self.ns['sbml'] + 'listOfUnitDefinitions')
         if x:
-            self._parse_units(model, comp, x)
+            self._parse_units(model, comp, x, self.ns['sbml'])
```
Comment (MichaelClerx) [inline myokit/formats/sbml/_importer.py:138]: Given that it's a property, I guess we don't strictly need to pass `self.ns['sbml']` here?

File: myokit/formats/sbml/_importer.py [hunk 8]
```diff
@@ -96,53 +102,76 @@ def model(self, path):
...
         if x:
-            self._parse_rules(model, comp, refs, x)
+            self._parse_rules(model, comp, refs, x, self.ns['sbml'])
 
         # Parse extra initial assignments
-        x = dom_child(xmodel, 'listOfInitialAssignments')
+        x = xmodel.find(self.ns['sbml'] + 'listOfInitialAssignments')
         if x:
-            self._parse_initial_assignments(model, comp, refs, x)
+            self._parse_initial_assignments(model,
```
Comment (MichaelClerx) [inline myokit/formats/sbml/_importer.py:169]: You can also just write
```
self._parse_initial_assignments(
    model, comp, refs, x, self.ns['sbml'])
```
I find that's usually more readable :D

File: myokit/formats/sbml/_importer.py [hunk 9]
```diff
@@ -159,42 +188,101 @@ def model(self, path):
         # Return finished model
         return model
 
-    def _parse_initial_assignments(self, model, comp, refs, node):
+    def _get_namespaces(self, sbml_version: str, log):
```
Comment (MichaelClerx) [inline myokit/formats/sbml/_importer.py:191]: Shouldn't we be getting the namespace from the file, instead of getting it from the version attributes?
Or maybe just check that the file combines the two correctly and complain if it says it's version 4 but uses ns for version 3?

File: myokit/formats/sbml/_importer.py [hunk 10]
```diff
@@ -159,42 +188,101 @@ def model(self, path):
...
+        """
+        Defines namespaces for supported SBML versions. Supported versions are:
+        level 2, version 3, 4, 5.
+        """
+        ns = dict()
+        if sbml_version == "{http://www.sbml.org/sbml/level2/version3}":
+            # SBML
+            ns['sbml'] = "{http://www.sbml.org/sbml/level2/version3}"
+            # MathML
+            ns['mathml'] = "{http://www.w3.org/1998/Math/MathML}"
```
Comment (MichaelClerx) [inline myokit/formats/sbml/_importer.py:201]: Does this change in past or future versions? Otherwise can just set once for all versions, I think

File: myokit/formats/sbml/_importer.py [hunk 11]
```diff
@@ -159,42 +188,101 @@ def model(self, path):
...
+                     % sbml_version + ' may not be imported correctly.'
+                     )
+
+        return ns
+
+    def _get_sbml_version(self, root):
+        """
+        Returns the SBML version of the file.
+        """
+        m = re.match(r'\{.*\}', root.tag)
```
Comment (MichaelClerx) [inline myokit/formats/sbml/_importer.py:239]: What's happening here? Can we not get the version from either the namespace of the root element, or from the attributes of the root element?

File: myokit/formats/sbml/_importer.py [hunk 12]
```diff
@@ -159,42 +188,101 @@ def model(self, path):
...
-
-    def _parse_parameters(self, model, comp, refs, node):
+    def _parse_parameters(self, model, comp, refs, node, ns):
         """
         Parses parameters
         """
-        node = dom_child(node, 'parameter')
-        while node:
+        nodes = node.findall(ns + 'parameter')
+        for node in nodes:
```
Comment (MichaelClerx) [inline myokit/formats/sbml/_importer.py:283]: Since `nodes` is never used again, you could also just write `for node in node.findall(..)

File: myokit/formats/sbml/_importer.py [hunk 13]
```diff
@@ -203,89 +291,102 @@ def _parse_parameters(self, model, comp, refs, node):
             else:
                 # Create variable
                 unit = None
-                if node.hasAttribute('units'):
-                    foreign_unit = node.getAttribute('units')
+                if node.get('units'):
```
Comment (MichaelClerx) [inline myokit/formats/sbml/_importer.py:294]: Not sure you need two lookups here!

Comment (MichaelClerx) [reply myokit/formats/sbml/_importer.py:294]: (The old minidom syntax didn't give a nicer way to do this)

File: myokit/formats/sbml/_importer.py [hunk 14]
```diff
@@ -203,89 +291,102 @@ def _parse_parameters(self, model, comp, refs, node):
...
-            node = dom_next(node, 'assignmentRule')
 
         # Create variables with rate rules (states)
-        node = dom_child(parent, 'rateRule')
-        while node:
+        nodes = parent.findall(ns + 'rateRule')
+        for node in nodes:
             var = self._convert_name(
-                str(node.getAttribute('variable')).strip())
+                str(node.get('variable')).strip())
```
Comment (MichaelClerx) [inline myokit/formats/sbml/_importer.py:343]: What happens if 'variable' doesn't exit?

File: myokit/formats/sbml/_importer.py [hunk 15]
```diff
@@ -203,89 +291,102 @@ def _parse_parameters(self, model, comp, refs, node):
...
-            while node2:
-                kind = str(node2.getAttribute('kind')).strip()
+            units = node.find(ns + 'listOfUnits')
+            units = units.findall(ns + 'unit')
+            for node2 in units:
+                kind = str(node2.get('kind')).strip()
                 u2 = self._convert_unit(kind)
-                if node2.hasAttribute('multiplier'):
-                    m = float(node2.getAttribute('multiplier'))
+                if node2.get('multiplier'):
```
Comment (MichaelClerx) [inline myokit/formats/sbml/_importer.py:379]: This looks like it could be more efficient?
What does node2.get('multiplier') do if multiplier isn't set?

File: myokit/tests/data/formats/sbml/TwoComp.xml [hunk 16]
```diff
@@ -0,0 +1,189 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (MichaelClerx) [inline myokit/tests/data/formats/sbml/TwoComp.xml:1]: If we're not using this one yet, I'd prefer it didn't come in until a future PR!

File: myokit/tests/test_formats_importers.py [hunk 17]
```diff
@@ -50,6 +50,12 @@ def test_unknown(self):
 
 class SBMLTest(unittest.TestCase):
     """ Test SBML import. """
+    # Test Case I: Hodgkin Huxley
+    i = formats.importer('sbml')
+    HoHu = i.model(os.path.join(
```
Comment (MichaelClerx) [inline myokit/tests/test_formats_importers.py:55]: Could we call this something in camel case?

File: myokit/tests/test_formats_importers.py [hunk 18]
```diff
@@ -50,6 +50,12 @@ def test_unknown(self):
...
+        DIR_FORMATS,
+        'sbml',
+        'HodgkinHuxley.xml'))
```
Comment (MichaelClerx) [inline myokit/tests/test_formats_importers.py:58]: Also, there's a `setUpClass` method you can use to initialise a unit test, instead of declaring class properties like this. See e.g. https://github.com/MichaelClerx/myokit/blob/master/myokit/tests/test_simulation_cvode.py#L34

Comment (MichaelClerx) [reply myokit/tests/test_formats_importers.py:58]: We can make multiple Test Classes for multiple models later, if you like

File: myokit/tests/test_formats_importers.py [hunk 19]
```diff
@@ -82,6 +88,145 @@ def sbml(fname):
...
+            'beta_m',
+            'alpha_h',
+            'beta_h',
+            'alpha_n',
+            'beta_n'
+        ]
+
+        # test whether parameters are in myokit model
+        for param in parameters:
+            assert self.HoHu.has_variable('sbml.' + param)
```
Comment (MichaelClerx) [inline myokit/tests/test_formats_importers.py:126]: I guess you've been using `pytest` :D 

In Myokit we're using plain `unittest` https://docs.python.org/3/library/unittest.html so this becomes `self.assertTrue(self.model.has_variable(..))`

File: myokit/tests/test_formats_importers.py [hunk 20]
```diff
@@ -82,6 +88,145 @@ def sbml(fname):
...
+            'alpha_h': None,
+            'beta_h': None,
+            'alpha_n': None,
+            'beta_n': None
+        }
+
+        # test whether parameters have correct units
+        for param in param_unit_dict:
+            unit = param_unit_dict[param]
+            assert unit == self.HoHu.get('sbml.' + param).unit()
```
Comment (MichaelClerx) [inline myokit/tests/test_formats_importers.py:164]: self.assertEqual(

File: myokit/tests/test_formats_importers.py [hunk 21]
```diff
@@ -82,6 +88,145 @@ def sbml(fname):
...
+            'g_L': 0.3,
+            'E_Na': -190,
+            'E_K': -63,
+            'E_L': -85.613
+        }
+
+        # test whether parameters have correct initial values
+        for param in param_value_dict:
+            value = param_value_dict[param]
+            assert value == self.HoHu.get('sbml.' + param).value()
```
Comment (MichaelClerx) [inline myokit/tests/test_formats_importers.py:184]: One downside of doing it with lists and dicts is that if it fails you don't get nice output saying which entry fails, it'll just tell you line 184 failed, but that's not top priority for now :-)

**Comment by MichaelClerx** [comment]
One thing I'd suggest you start with:

1. undo the changes to the OpenCL files in this branch
2. `git merge develop` to get the new fixes for OpenCL on os/x

File: myokit/formats/sbml/_importer.py (hunk 9 above)
Comment (MichaelClerx) [inline myokit/formats/sbml/_importer.py:191]: Myokit is used on Python 2 and Python 3, so `sbml_version : str` will have to become just `sbml_version`, I'm afraid

**Comment by DavAug** [comment]
I pulled the latest version of the develop branch, reverted the changes in the opencl.py, openclsim.py and fiber_tissue.py file and then merged the develop branch into sbml-etree. This reintroduced the failures of the tests again, see example below.

Was this expected?

```
======================================================================
ERROR: test_simulation_p (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: test_simulation_p
Traceback (most recent call last):
  File "/usr/local/Cellar/python/3.7.5/Frameworks/Python.framework/Versions/3.7/lib/python3.7/unittest/loader.py", line 436, in _find_test_path
    module = self._get_module_from_name(name)
  File "/usr/local/Cellar/python/3.7.5/Frameworks/Python.framework/Versions/3.7/lib/python3.7/unittest/loader.py", line 377, in _get_module_from_name
    __import__(name)
  File "/Users/david/Workspace/sabs-r3/myokit/myokit/tests/test_simulation_p.py", line 17, in <module>
    from shared import DIR_DATA, CancellingReporter
  File "/Users/david/Workspace/sabs-r3/myokit/myokit/tests/shared.py", line 33, in <module>
    OpenCL_FOUND = myokit.OpenCL.supported()
  File "/Users/david/Workspace/sabs-r3/myokit/myokit/_sim/opencl.py", line 206, in supported
    OpenCL._get_instance()
  File "/Users/david/Workspace/sabs-r3/myokit/myokit/_sim/opencl.py", line 90, in _get_instance
    OpenCL()
  File "/Users/david/Workspace/sabs-r3/myokit/myokit/_sim/opencl.py", line 65, in __init__
    mname, fname, args, libs, libd, incd)
  File "/Users/david/Workspace/sabs-r3/myokit/myokit/_sim/__init__.py", line 197, in _compile
    return load_module(name, d_modul)
  File "/Users/david/Workspace/sabs-r3/myokit/myokit/_sim/__init__.py", line 54, in load_module
    module = importlib.util.module_from_spec(spec)
ImportError: dlopen(/var/folders/8w/b9lb3wb92c1fscyr5tc8nswh0000gp/T/tmphwfhy8yimyokit/module/myokit_opencl_info_36.cpython-37m-darwin.so, 2): Symbol not found: _clGetDeviceIDs
  Referenced from: /var/folders/8w/b9lb3wb92c1fscyr5tc8nswh0000gp/T/tmphwfhy8yimyokit/module/myokit_opencl_info_36.cpython-37m-darwin.so
  Expected in: flat namespace
 in /var/folders/8w/b9lb3wb92c1fscyr5tc8nswh0000gp/T/tmphwfhy8yimyokit/module/myokit_opencl_info_36.cpython-37m-darwin.so


----------------------------------------------------------------------
```

File: myokit/formats/sbml/_importer.py (hunk 1 above)
Comment (DavAug) [reply myokit/formats/sbml/_importer.py:143]: I was planning on implementing the time read in properly in the next pull request. Apparently it has changed how the SBML versions deal with time, and dependent on the version they assume different base units and different identifiers. 

Should I then remove the time binding for now and incorporate and introduce it with the next pull request?

File: myokit/_sim/fiber_tissue.py (hunk 2 above)
Comment (DavAug) [reply myokit/_sim/fiber_tissue.py]: reverted and merged the develop branch, bit this didn't fix the bug, see comment below.

**Comment by MichaelClerx** [comment]
@DavAug I accidentally sent a fix to `master` instead of `develop` :man_facepalming: 

Have stuck the fix on develop now too, could you try merging in develop again for me? Sorry!

File: myokit/formats/sbml/_importer.py (hunk 3 above)
Comment (DavAug) [reply myokit/formats/sbml/_importer.py:71]: Will do, once we decided what to do with the time binding, see comment above.

File: myokit/formats/sbml/_importer.py (hunk 4 above)
Comment (DavAug) [reply myokit/formats/sbml/_importer.py:152]: Done

File: myokit/formats/sbml/_importer.py (hunk 5 above)
Comment (DavAug) [reply myokit/formats/sbml/_importer.py:147]: Is also more elegant. Done.

File: myokit/formats/sbml/_importer.py (hunk 6 above)
Comment (DavAug) [reply myokit/formats/sbml/_importer.py:105]: Should we push this issue to the next pull request? I will need to read more closely through the SBML specifications anyway to deal with time, compartments, species etc. and then I can make sure that the file obeys the prescribed format.

File: myokit/formats/sbml/_importer.py (hunk 7 above)
Comment (DavAug) [reply myokit/formats/sbml/_importer.py:138]: Agreed. ns wasn't a property before and I forgot to change the input. Done.

File: myokit/formats/sbml/_importer.py (hunk 8 above)
Comment (DavAug) [reply myokit/formats/sbml/_importer.py:169]: Ok. Done.

File: myokit/formats/sbml/_importer.py (hunk 10 above)
Comment (DavAug) [reply myokit/formats/sbml/_importer.py:201]: This won't change, but other properties such as base unit definitions, and other namespaces. For now we could, however, just add the 'mathml' namespace irrespective of the sbml version

Comment (DavAug) [reply myokit/formats/sbml/_importer.py:201]: Done.

File: myokit/formats/sbml/_importer.py (hunk 11 above)
Comment (DavAug) [reply myokit/formats/sbml/_importer.py:239]: There doesn't seem to be a native function in etree to access the namespace of any variable. So what this does is it gets the tag of the root, i.e. the name/ id, and straps off it's namespace, which is the SBML version.

I haven't found a better way to do this.

File: myokit/formats/sbml/_importer.py (hunk 12 above)
Comment (DavAug) [reply myokit/formats/sbml/_importer.py:283]: Agreed, and changed everywhere.

File: myokit/formats/sbml/_importer.py (hunk 13 above)
Comment (DavAug) [reply myokit/formats/sbml/_importer.py:294]: Agreed. Done.

File: myokit/formats/sbml/_importer.py (hunk 15 above)
Comment (DavAug) [reply myokit/formats/sbml/_importer.py:379]: node.get() returns None if multiplier does not exist, so it wouldn't do anything.

Comment (DavAug) [reply myokit/formats/sbml/_importer.py:379]: I'm not 100 percent sure how myokit units behave, but would something like this produce the same 
```
kind = str(node2.get('kind')).strip()
u2 = self._convert_unit(kind)
if node2.get('multiplier'):
      u2 *= float(node2.get('multiplier'))
 if node2.get('scale'):
       u2 *= 10 ** float(node2.get('scale'))
 if node2.get('exponent'):
        u2 **= float(node2.get('exponent'))
 unit *= u2
```

And what happens anyway if a unit has several `kind`. I wanted to clarify these questions for the next pull request and just kept your structure and simply replaced minidom by etree everywhere.

File: myokit/tests/data/formats/sbml/TwoComp.xml (hunk 16 above)
Comment (DavAug) [reply myokit/tests/data/formats/sbml/TwoComp.xml:1]: Ok, I removed it.

File: myokit/tests/test_formats_importers.py (hunk 17 above)
Comment (DavAug) [reply myokit/tests/test_formats_importers.py:55]: Yep, done.

File: myokit/tests/test_formats_importers.py (hunk 18 above)
Comment (DavAug) [reply myokit/tests/test_formats_importers.py:58]: Ok, sounds good! I used the setUpClass method now.

File: myokit/tests/test_formats_importers.py (hunk 19 above)
Comment (DavAug) [reply myokit/tests/test_formats_importers.py:126]: Ok, changed.

File: myokit/tests/test_formats_importers.py (hunk 20 above)
Comment (DavAug) [reply myokit/tests/test_formats_importers.py:164]: Done

File: myokit/tests/test_formats_importers.py (hunk 21 above)
Comment (DavAug) [reply myokit/tests/test_formats_importers.py:184]: Yes, that's something I had to deal with a couple times now. How would you do it then? write out the asserts for each parameter explicitly? I'm happy to adapt for the next pul request.

File: myokit/formats/sbml/_importer.py (hunk 14 above)
Comment (DavAug) [reply myokit/formats/sbml/_importer.py:311]: At this point it is assumed that the file adheres to the SBML file format. It would make sense to have a rateRule without defining the lhs, i.e. the variable. But as mentioned above, I basically just kept your code and replaced mindom by etree. 

I will look thorugh the exact specifications in the next PR and make sure that only files following those will be read in, if that's ok.

File: myokit/formats/sbml/_importer.py (hunk 9 above)
Comment (DavAug) [reply myokit/formats/sbml/_importer.py]: I guess that would make sense! I have, however, not found a native function in etree to access the namespaces in a file. What we could do is look at the SBML definitions, where namespaces should be expected to occur in the file, and then look at all those places. But we probably would have to start branching out from the root, because I don't global searches are possible.

What would you prefer?

**Comment by DavAug** [comment]
Haha ok thats strange. Your develop branch passes but when I merge it into my branch it fails?

**Comment by DavAug** [comment]
Also locally on my OS the errors remained the same.

```
======================================================================
ERROR: test_simulation_p (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: test_simulation_p
Traceback (most recent call last):
  File "/usr/local/Cellar/python/3.7.5/Frameworks/Python.framework/Versions/3.7/lib/python3.7/unittest/loader.py", line 436, in _find_test_path
    module = self._get_module_from_name(name)
  File "/usr/local/Cellar/python/3.7.5/Frameworks/Python.framework/Versions/3.7/lib/python3.7/unittest/loader.py", line 377, in _get_module_from_name
    __import__(name)
  File "/Users/david/Workspace/sabs-r3/myokit/myokit/tests/test_simulation_p.py", line 17, in <module>
    from shared import DIR_DATA, CancellingReporter
  File "/Users/david/Workspace/sabs-r3/myokit/myokit/tests/shared.py", line 33, in <module>
    OpenCL_FOUND = myokit.OpenCL.supported()
  File "/Users/david/Workspace/sabs-r3/myokit/myokit/_sim/opencl.py", line 209, in supported
    OpenCL._get_instance()
  File "/Users/david/Workspace/sabs-r3/myokit/myokit/_sim/opencl.py", line 93, in _get_instance
    OpenCL()
  File "/Users/david/Workspace/sabs-r3/myokit/myokit/_sim/opencl.py", line 68, in __init__
    mname, fname, args, libs, libd, incd, larg=flags)
  File "/Users/david/Workspace/sabs-r3/myokit/myokit/_sim/__init__.py", line 199, in _compile
    return load_module(name, d_modul)
  File "/Users/david/Workspace/sabs-r3/myokit/myokit/_sim/__init__.py", line 54, in load_module
    module = importlib.util.module_from_spec(spec)
ImportError: dlopen(/var/folders/8w/b9lb3wb92c1fscyr5tc8nswh0000gp/T/tmpryt3u3iumyokit/module/myokit_opencl_info_36.cpython-37m-darwin.so, 2): Symbol not found: _clGetDeviceIDs
  Referenced from: /var/folders/8w/b9lb3wb92c1fscyr5tc8nswh0000gp/T/tmpryt3u3iumyokit/module/myokit_opencl_info_36.cpython-37m-darwin.so
  Expected in: flat namespace
 in /var/folders/8w/b9lb3wb92c1fscyr5tc8nswh0000gp/T/tmpryt3u3iumyokit/module/myokit_opencl_info_36.cpython-37m-darwin.so


----------------------------------------------------------------------
Ran 391 tests in 5.650s

FAILED (errors=36)
FAILED
```

**Comment by DavAug** [comment]
I just doubled checked. Also the develop branch does not pass the tests on my machine.

**Comment by MichaelClerx** [comment]
Hmmmm, that's a shame. Well, we can ignore the OpenCL errors for now, in this branch I guess

File: myokit/formats/sbml/_importer.py (hunk 1 above)
Comment (MichaelClerx) [reply myokit/formats/sbml/_importer.py:143]: No it's fine to keep it in there, I just think we should have some strategy in case the model contains a variable called "time", because that will cause this line to fail. There's an `add_variable_allow_renaming(name)` method you could use instead, for example?

File: myokit/formats/sbml/_importer.py (hunk 6 above)
Comment (MichaelClerx) [reply myokit/formats/sbml/_importer.py:105]: Yeah it's fine. I'm happy with an importer that just works most of the time, for valid files. There's plenty of validation tools etc. out there already without us adding one :-)

File: myokit/formats/sbml/_importer.py (hunk 11 above)
Comment (MichaelClerx) [reply myokit/formats/sbml/_importer.py:216]: Oh ok! I see what this is doing now.
There's a method `mxml.split` that splits off the namespace in much the same way. Might be worth using that instead. (Not "better", just means we're using a single method consistently and can fix bugs more easily etc)

File: myokit/formats/sbml/_importer.py (hunk 14 above)
Comment (MichaelClerx) [reply myokit/formats/sbml/_importer.py:311]: Cool!

That's ok as long as we "fail gracefully".
For example, I think `get()` returns `None` if an attribute doesn't exist, so if we do `str(get(..))` we'd get a string 'None' which might not be what we want

File: myokit/formats/sbml/_importer.py (hunk 15 above)
Comment (MichaelClerx) [reply myokit/formats/sbml/_importer.py:347]: Fair enough. I'm just not sure if get() is an expensive operation. Probably not?
I was thinking `x = get(); if x: ...` would give us a single get
Up to you :D

File: myokit/formats/sbml/_importer.py (hunk 9 above)
Comment (MichaelClerx) [reply myokit/formats/sbml/_importer.py]: I think you just need to get it from each element separately, so like you're doing with the regex or using `mxml.split()`.
So I could write
```
<model xmlns="namespace">
  <component />
</model>
```
or
```
<sbml:model xmlns:sbml="namespace">
  <sbml:component />
</sbml:model>
```
or
```
<model xmlns="other namespace">
  <component xmlns="namespace" />
</model>
```
and in all of those the `component` tag would be a "{namespace}component".

So no file-wide namespaces, each element has one!

File: myokit/tests/test_formats_importers.py (hunk 21 above)
Comment (MichaelClerx) [reply myokit/tests/test_formats_importers.py]: Exactly. It's not super important but the best way is to write explicit asserts I'm afraid. Maybe I should consider switching to pytest :D

File: myokit/formats/sbml/_importer.py [hunk 22]
```diff
@@ -96,53 +102,72 @@ def model(self, path):
         refs = {}
 
         # Handle notes, if given
-        x = dom_child(xmodel, 'notes')
+        x = xmodel.find(self.ns['sbml'] + 'notes')
         if x:
             log.log('Converting <model> notes to ascii')
-            model.meta['desc'] = html2ascii(x.toxml(), width=75)
+            model.meta['desc'] = html2ascii(ET.tostring(x, encoding='unicode'),
```
Comment (MichaelClerx) [inline myokit/formats/sbml/_importer.py:108]: I'm guessing that should be encoding='utf-8' ?

File: myokit/formats/sbml/_importer.py (hunk 15 above)
Comment (MichaelClerx) [reply myokit/formats/sbml/_importer.py:347]: Actually, are there default options specified for these attributes? `.get()` has a `default` option, so you might be able to do something like
`u2 *= float(node2.get('multiplier', default=1))`
etc.

**Comment by MichaelClerx** [comment]
Thanks @DavAug ! I think once you changed the `encoding` it should pass in python 2.7 as well and then we're pretty much done!

File: myokit/formats/sbml/_importer.py (hunk 22 above)
Comment (DavAug) [reply myokit/formats/sbml/_importer.py:108]: I chose unicode because any other encoding returns a bit string which then has to be converted back. But I can change that, no prob.

Comment (MichaelClerx) [reply myokit/formats/sbml/_importer.py:108]: It seems to be failing in Python 2.7

Comment (MichaelClerx) [reply myokit/formats/sbml/_importer.py:108]: utf-8 should be unicode too, so not sure what's going on there

File: myokit/formats/sbml/_importer.py (hunk 1 above)
Comment (DavAug) [reply myokit/formats/sbml/_importer.py:143]: OK. Done.

File: myokit/formats/sbml/_importer.py (hunk 3 above)
Comment (DavAug) [reply myokit/formats/sbml/_importer.py:71]: Done.

File: myokit/formats/sbml/_importer.py (hunk 9 above)
Comment (DavAug) [reply myokit/formats/sbml/_importer.py]: I am using split to access the namespace. I would keep the namespace specification in the beginning for now and the maths ns for the mathy bits for now, as otherwise I will have to prepare a comprehensive list of all the namespaces defined within an SBML version and I would rather do that for the next PR, where the different namespaces may have more relevance. At the moment it seems, we are mostly interested in the maths namespace, so we can translate the equations properly and the gobal version ns to find the relevant nodes.

File: myokit/formats/sbml/_importer.py (hunk 11 above)
Comment (DavAug) [reply myokit/formats/sbml/_importer.py:216]: Done

File: myokit/formats/sbml/_importer.py (hunk 14 above)
Comment (DavAug) [reply myokit/formats/sbml/_importer.py:311]: Done

File: myokit/formats/sbml/_importer.py (hunk 15 above)
Comment (DavAug) [reply myokit/formats/sbml/_importer.py:347]: Done

File: myokit/tests/test_formats_importers.py (hunk 21 above)
Comment (DavAug) [reply myokit/tests/test_formats_importers.py]: Ok I will leave it for now, and change in the next PR

File: myokit/formats/sbml/_importer.py [hunk 23]
```diff
@@ -67,21 +66,34 @@ def _convert_unit(self, unit):
     def info(self):
         return info
 
-    def model(self, path):
+    def model(self, path, bind_time=True):
+        """
+        Returns a :class:myokit.Model based on the SBML file provided.
 
+        Arguments:
+            path -- Path to SBML file.
```
Comment (MichaelClerx) [inline myokit/formats/sbml/_importer.py:74]: Please see e.g. [https://github.com/MichaelClerx/myokit/blob/master/myokit/_model_api.py#L319](here) for the syntax we're using in Myokit.
